Add Notification Center

nwmac opened the pull request:
### Summary
Fixes #14007 

### Occurred changes and/or fixed issues

This PR adds support for the new Notification Center UI that is shown in the top-right of the UI.

Currently this is only wired into 2 places:

- Release notes - the notification to read the release notes for a release notes has been moved from the home page banner into the notification center
- Growls - all growls will add a corresponding notification to the notification center

### Technical notes summary

To ensure backwards compatibility with UI Extensions, we leave the existing `growl` store in place and add a new `notification` store. This should be the considered the go-forward API that developers should use and the `growl` store should be considered deprecated.

Note that for now, developers should use the notification store directly - we will add an API for this later as part of introducing APIs for extensions and developers.

Of note:

- e2e tests for release notes have been updated to cover the use of the notification center
- `GrowlManager` has been updated to fix a bug where one or more growls could get stuck on screen. The button that would show to `Clear all` has been removed, since growls will appear in the notification center. The growl notifications now also show below the header, rather than at the top of the page, so that they do not overlap on the new notification center icon
- Growl store - We will only show 5 growls at most - since growls are also in the notification center, if 5 are on screen and a new one appears, the oldest will be removed - they are always available in the notification center.
- Growl store - Growls now send notifications. A new action is added, `notification`, since notifications will also show growls for `success`, `warning` and `error` growls (not info and others). This ensures we do not get any circular issues.
- Growl store - errors do not override the timeout to be shown indefinitely - they will use the 5 second default and auto disappear like the others, since the user will be able to access them via the notification center
- Growl store - closing a growl that was creating from a notification will mark the associated notification as read
- Notification store - pretty straight-forward. Notifications are persisted in local storage in the browser. We keep at most 50 notifications and 14 days of notifications
- The notification center UI uses the new `RcDropDown` with custom items. A lot of effort has been put into making sure this is accessible.
- The notification center uses a focus trap for each item, so when the user clicks, they tab between the read/unread button and the primary and secondary buttons (if they are present) - the component manages this directly as the focus trap util did not seem to work for this use case
- Release notes will not check at startup if it needs to add a notification. It checks the user preference and uses a notification with a well-known identifier to ensure we only re-add the release notes notification for a release when necessary.

### Areas or cases that should be tested

e2e tests cover release notes.

We may want to do some manual testing of the UI and the keyboard navigation.

### Screenshot/Video

Example with some test notifications:

![image](https://github.com/user-attachments/assets/04e1c851-5713-4c20-a36a-b67f272709eb)

Notifications all clear:

![image](https://github.com/user-attachments/assets/a8461aa2-0680-4e92-91a7-eae73f738f81)

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [x] The PR template has been filled out
- [x] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes
- [x] The PR has been reviewed in terms of Accessibility

rak-phillip commented on shell/store/notifications.ts:
```diff
@@ -0,0 +1,272 @@
+  window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(state.notifications));
+}
+
+export const state = function(a: any): NotificationsStore {
+  let notifications = [];
+
+  try {
+    notifications = JSON.parse(window.localStorage.getItem(LOCAL_STORAGE_KEY) || '[]');
+  } catch (e) {
+    console.error('Unable to read notifications from local storage', e); // eslint-disable-line no-console
+  }
+
+  // Expire old notifications
+  const now = new Date();
+
+  notifications = notifications.filter((n: StoredNotification) => {
+    // Try ... catch in case the date parsing fails
+    try {
+      const created = new Date(n.created);
+      const diff = (now.getTime() - created.getTime()) / 1000; // Diff in seconds
+
+      return diff < EXPIRY;
+    } catch (e) {}
+
+    return true;
+  });
+
+  return { notifications };
+};
```
Similar to a data prop in a vue component, a store's state should avoid complex logic to focus on declaring reactive state. Would this logic initialization logic better fit in the `init()` action?

nwmac replied:
Good idea - done

rak-phillip reviewed (changes requested):
What's here looks really good - It's good to see that the RcDropdown component can be extended to support additional use cases.

The notification dot looks a little wrong in dark mode, almost as if slightly squared and has a few white pixels bleeding through at the border? I think it's more pronounced when it's in the checked state:

![image](https://github.com/user-attachments/assets/143f0cd3-6e54-42c7-ae73-0a348a3d9436)

![image](https://github.com/user-attachments/assets/ce4208ac-404d-48dd-8a9b-8f780329e0db)

I'm not sure if I like having `escape` be the only hatch to exit the focus trap. I feel that being able to press up/down on they keyboard to exit the trap and focus the next item would feel more natural. I think that we should let others play around with the existing implementation and collect feedback on how they feel about the focus trap interaction.

nwmac commented on shell/store/notifications.ts:
```diff
@@ -0,0 +1,272 @@
+const LOCAL_STORAGE_KEY = '#notifications';
+
+// Expire in seconds (14 days)
+const EXPIRY = 14 * 24 * 60 * 60;
```
We don't currently bring in the `duration` plugin for `dayjs`, which I think is what we'd need here. I had a look, but it was more change than I expected - can we do this later?

nwmac commented on shell/store/notifications.ts:
```diff
@@ -0,0 +1,272 @@
+}
+
+export const state = function(a: any): NotificationsStore {
+  let notifications = [];
```
Done

nwmac commented:
@rak-phillip I think I've addressed most of your feedback - I did create a follow-on issue for a couple of things - https://github.com/rancher/dashboard/issues/14512 - I'd like to get this in, so it can be wired in to the registration work.

rak-phillip reviewed (commented):
I think that this is looking good. I left a comment for a potential improvement of scrolling behavior.

The tab behavior is also less than ideal when attempting to tab away while the notification center is open. We should focus on the next element in the tab order, but it appears that an element in the notification center receives focus before the menu is closed. This isn't too disruptive, but it leaves us in a confusing state. You can see what I'm talking about if you compare the tab behavior of the page actions menu when it is open and compare it to the notification center.

Everything else looks good at this point.

nwmac commented on shell/components/nav/NotificationCenter/index.vue:
```diff
@@ -0,0 +1,137 @@
+<script setup lang="ts">
```
Yeah, I know. Discussed with UX and said it was okay as is.

I was looking at adding something in the User Preferences - maybe I can do that as part of the follow-on ticket.

nwmac commented on shell/components/nav/NotificationCenter/Notification.vue:
```diff
@@ -0,0 +1,446 @@
+</script>
+
+<template>
+  <div
+    ref="dropdownMenuItem"
+    dropdown-menu-item
+    tabindex="-1"
+    role="menuitem"
+    data-testid="notifications-center-item"
+    :aria-label="t('notificationCenter.ariaLabel', { title: item.title })"
+    @keydown.up.down.stop.prevent="handleKeydown"
+    @focus.stop="gotFocus"
+    @click.stop
+    @keydown.enter.space.stop="enterFocusTrap"
+  >
```
Nice. Thank you Phil. I have updated the PR with this change.

nwmac commented on docs/developer/notification-center.md:
```diff
@@ -0,0 +1,95 @@
+
+## Adding a Notification
+
+> Note: Types are defined in `shell/notifications.ts`
```
Moves the types

nwmac commented on shell/utils/release-notes.ts:
```diff
@@ -0,0 +1,47 @@
+        key:   READ_WHATS_NEW,
+        value: rancherVersion
+      },
+      primaryAction: {
```
Yes, we should

nwmac commented on docs/developer/notification-center.md:
```diff
@@ -0,0 +1,95 @@
+# Notification Center API
```
I have added a comment

nwmac commented on shell/components/nav/NotificationCenter/Notification.vue:
```diff
@@ -0,0 +1,446 @@
+<script setup lang="ts">
```
Would love this - I can't say I love the composition API

nwmac commented on shell/store/growl.js:
```diff
@@ -50,11 +63,29 @@ export const actions = {
+
+  async success({ commit, dispatch }, data) {
+    // Send a notification for the growl
+    const notification = await dispatch('notifications/fromGrowl', {
```
I've tuned those down to be `info` growls, rather than `success` - we can revisit, but this stops them also appearing in the notification center.

nwmac commented on shell/store/notifications.ts:
```diff
@@ -0,0 +1,287 @@
+import { randomStr } from '@shell/utils/string';
+
+const LOCAL_STORAGE_KEY = '#notifications';
```
Yeah, It doesn't matter - I thought i'd use that. I can remove it you like.

nwmac commented on shell/store/index.js:
```diff
@@ -881,6 +882,11 @@ export const actions = {
 
+    // Add the notification for the release notes
+    if (isRancher) {
+      await addReleaseNotesNotification(dispatch, getters);
```
That is what I am hoping

nwmac commented on docs/developer/notification-center.md:
```diff
@@ -0,0 +1,95 @@
+
+![Example Notification](./images/notification-example.png)
+
+Note that the `id` field is optional - if not specified, the notification will be assigned a unique identifier. Using a known identifier is only really required if you wish to be able to look up a notification by a known identifier OR you wish to be able yo update a notification after being added.
```
Done

nwmac commented on docs/developer/notification-center.md:
```diff
@@ -0,0 +1,95 @@
+
+![Example Notification](./images/notification-example.png)
+
+Note that the `id` field is optional - if not specified, the notification will be assigned a unique identifier. Using a known identifier is only really required if you wish to be able to look up a notification by a known identifier OR you wish to be able yo update a notification after being added.
```
Added a check for an existing notification with the same id

nwmac commented on shell/store/notifications.ts:
```diff
@@ -0,0 +1,287 @@
+    persist(state);
+  },
+
+  update(state: NotificationsStore, notification: Notification) {
```
Done

nwmac commented on shell/components/nav/NotificationCenter/Notification.vue:
```diff
@@ -0,0 +1,446 @@
+    role="menuitem"
+    data-testid="notifications-center-item"
+    :aria-label="t('notificationCenter.ariaLabel', { title: item.title })"
+    @keydown.up.down.stop="handleKeydown"
```
Added

nwmac commented on shell/components/nav/NotificationCenter/Notification.vue:
```diff
@@ -0,0 +1,446 @@
+  // Return focus to the outer notification div
+  dropdownMenuItem.value?.focus();
+
+  e.stopPropagation();
+  e.preventDefault();
```
Will leave for now

rak-phillip commented on docs/developer/notification-center.md:
```diff
@@ -0,0 +1,97 @@
+
+> &#x26a0;&#xfe0f; Documentation in this directory is intended for internal use only. Any information contained here is unsupported.
+
+> For now, developers can use the Notification Centre via the VueX store. In the future, we will add an API for this, that will be fully supported as part of the UI Extensions API.
```
nit: we switch between "Centre" and "Center". We should stick to one for consistency and use that throughout 🙂

nwmac replied:
Indeed - I try to use the US spelling, but my better conscience is struggling with it

rak-phillip reviewed (approved):
LGTM

I found an edge case where the notification header receives focus if the list is empty. I think that this is an issue with the Dropdown Menu implementation not respecting `tabindex=-1`.

![image](https://github.com/user-attachments/assets/4e247770-2175-4857-928d-776e25252414)
